builtin/logical/pki/cert_util.go

```diff
@@ -708,6 +732,13 @@ func generateCreationBundle(b *backend, data *dataBundle) error {
 			}
 		}
 
+		if data.csr != nil && len(data.role.AllowedSerialNumbers) > 0 {
+			ridSerialNumber = data.csr.Subject.SerialNumber
+		}
+		if ridSerialNumber == "" {
```

Comment:
I like the check that you added but think the order should be reversed -- API-specified values should override CSR values. Part of the reason is that a lot of people are not careful when constructing CSRs and may pick the same value over and over (e.g. just set something in a template) or may not use a sufficiently random source if they aren't intending for it to be something specific.